engines/scumm/he/moonbase/map_spiff.cpp

```diff
@@ -0,0 +1,623 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/debug.h"
+
+#include "scumm/he/moonbase/map_spiff.h"
+
+namespace Scumm {
+
+SpiffGenerator::SpiffGenerator(int seed) {
+	_seed = seed;
+}
+
+SpiffGenerator::~SpiffGenerator() {
+}
+
+MapFile *SpiffGenerator::generateMap(int water, int tileset, int mapSize, int energy, int terrain) {
+	totalMapSizeG = mapSize;
+	energyAmountG = (2 + energy) * totalMapSizeG * totalMapSizeG;
+
+	islandsFlagG = pickFrom2(0, 1, water - 4, (water >= 5)); // 1 is large islands, 2 is small
+	if (islandsFlagG) {
+		water -= 3;
+		energyAmountG = (int)(energyAmountG * (5 - islandsFlagG) / 6); // *2/3 or *1/2
+	}
+
+	waterAmountG = 4 * water;
+	cliffAmountG = 1 << terrain;
+	advancedMirrorOK_G = ((terrain > 1) && (water < 6)) || islandsFlagG;
+	terrainSeedFlagG = 2 * water - terrain;
+
+	int n = (int)(energyAmountG / 2700);
+	if (n > 12) {
+		n = 12;
+	}
+	if (n < 1) {
+		n = 1;
+	}
+	numPoolsG = spiffRand((int)(energyAmountG / 4000) + 1, n);
+	if (numPoolsG > 12) {
+		numPoolsG = 12;
+	}
+
+	generate();
+
+	// Populate MIF for map data generation:
+	MIF mif = MIF();
+
+	int levelMap[MAXELEVVAL];
+	levelMap[HIGH] = 2;
+	levelMap[MEDIUM] = 1;
+	levelMap[LOW] = 0;
+
+	mif.mapType = tileset;
+	Common::sprintf_s(mif.name, "Spiff %04X", (uint16)_seed);
+
+	mif.dimension = totalMapSizeG;
+
+	int y;
+	int x;
+	char t;
+	int XOffset = spiffRand(0, totalMapSizeG-1);
+	int YOffset = spiffRand(0, totalMapSizeG-1);
+	int newX;
+	int newY;
+
+	for (y = 0, newY = YOffset; y < totalMapSizeG; ++y, ++newY) {
+		for (x = 0, newX = XOffset; x < totalMapSizeG; ++x, ++newX) {
+			if (newX == totalMapSizeG)
+				newX = 0;
+			if (newY == totalMapSizeG)
+				newY = 0;
+			mif.cornerMap[newX][newY] = levelMap[mapCorner[x][y]];
+			switch (mapMiddle[x][y]) {
+			case HUB:
+				t = -1;
+				break;
+			case SMALLPOOL:
+				t = 'S';
+				break;
+			case MEDIUMPOOL:
+				t = 'M';
+				break;
+			case LARGEPOOLBOTTOM:
+				t = 'L';
+				break;
+			case WATER:
+				t = 'W';
+				break;
+			case UNASSIGNED:
+			case LARGEPOOLTOP:
+				t = '.';
+				break;
+			default:
+				t = '?';
+			}
+			mif.centerMap[newX][newY] = t;
+		}
+	}
+
+	// Generate new map:
+	MapFile *map = new MapFile();
+	mif.generateMap(map);
+
+	return map;
+}
+
+float SpiffGenerator::getRandomFloat() {
+	// This is the exact linear congruential generator
+	// algorithm used on MSVCRT (Windows Visual C++ Runtime), with
+	// RAND_MAX being 0x7fff (32767).  This is implemented here
+	// to match the RNG between the original Moonbase Console
+	// program and ScummVM.
+	//
+	// (Common::RandomSource uses Xorshift and uses unsigned
+	// integers compared to MSVCRT's rand)
+	_seed = _seed * 214013 + 2531011;
+	return (float)((_seed >> 16) & 0x7fff) / 32767;
```

Comment:
Why is it `float` when all operations are integer?

Reply:
As the comment clearly states, this function exactly mimics the MSVCRT's `rand()` function.  It is necessary because `Common::RandomSource` will break the Spiff generator.

If this is of any serious concern, what do you think we should do?